src/debug/ee/debugger.h

```diff
@@ -1072,6 +1072,138 @@ class DebuggerMethodInfo
+//
+// Each DebuggerHeapAllocatorPage is divided into 64 of these chunks.
+// The first chunk is a BookkeepingChunk used for bookkeeping information
+// for the page, and the remaining ones are DataChunks and are handed out
+// by the allocator when it allocates memory.
+// ------------------------------------------------------------------------ */
+union DebuggerHeapAllocatorChunk {
+
+    struct DataChunk
+    {
+        char data[48];
+
+        DebuggerHeapAllocatorPage *startOfPage;
+
+        // The chunk number within the page.
+        uint8_t chunkNumber;
+
+    } d;
+
+    struct BookkeepingChunk
+    {
+        DebuggerHeapAllocatorPage *nextPage;
+
+        uint64_t pageOccupancy;
+
+        char padding[48];
+
+    } b;
+
+} __attribute__((__aligned__(64)));
+
+// ------------------------------------------------------------------------ */
+// DebuggerHeapAllocatorPage
+//
+// We allocate the size of DebuggerHeapAllocatorPage each time we need
+// more memory and divide each page into DebuggerHeapAllocatorChunks for
+// use. The pages are self describing; the first chunk contains information
+// about which of the other chunks are used/free as well as a pointer to 
+// the next page.
+// ------------------------------------------------------------------------ */
+struct DebuggerHeapAllocatorPage
+{
+    inline DebuggerHeapAllocatorPage* GetNextPage()
+    {
+        return chunks[0].b.nextPage;
+    }
+
+    inline void SetNextPage(DebuggerHeapAllocatorPage* nextPage)
+    {
+        chunks[0].b.nextPage = nextPage;
+    }
+
+    inline uint64_t GetPageOccupancy()
+    {
+        return chunks[0].b.pageOccupancy;
+    }
+
+    inline void SetPageOccupancy(uint64_t newOccupancy)
+    {
+        // Can't unset first bit of occupancy!
+        ASSERT((newOccupancy & 0x8000000000000000) != 0);
+
+        chunks[0].b.pageOccupancy = newOccupancy;
+    }
+
+    inline void* GetPointerToChunk(int chunkNum)
+    {
+        return (char*)this + chunkNum * sizeof(DebuggerHeapAllocatorChunk);
+    }
+
+    DebuggerHeapAllocatorPage()
+    {
+        SetPageOccupancy(0x8000000000000000); // only the first bit is set.
+        for (uint8_t i = 1; i < sizeof(chunks)/sizeof(chunks[0]); i++)
+        {
+            ASSERT(i != 0);
+            chunks[i].d.startOfPage = this;
+            chunks[i].d.chunkNumber = i;
+        }
+    }
+
+private:
+    DebuggerHeapAllocatorChunk chunks[64];
+
+} __attribute__((__aligned__(4096)));
+
+// ------------------------------------------------------------------------ */
+// DebuggerHeapAllocationManager class
+// Handles allocation and freeing (and all necessary bookkeeping) for 
+// executable memory that the DebuggerHeap class needs. This is especially
+// useful on systems (like SELinux) where having executable code on the
+// heap is explicity disallowed for security reasons.
+// ------------------------------------------------------------------------ */
+
+class DebuggerHeapAllocationManager
```

Comment:
As far as naming I think "Executable" should be in the name and we probably don't need "Manager" so something like DebuggerExecutableHeap.